gateway/testutil.go

```diff
@@ -1074,71 +1073,74 @@ func (s *Test) BootstrapGw(ctx context.Context, cancelFn context.CancelFunc, gen
 		}
 	}()
 
-	s.Gw.keyGen = DefaultKeyGenerator{Gw: s.Gw}
-	s.Gw.CoProcessInit()
-	s.Gw.afterConfSetup()
+	gw.keyGen = DefaultKeyGenerator{Gw: gw}
+	gw.CoProcessInit()
+	gw.afterConfSetup()
 
 	defaultTestConfig = gwConfig
-	s.Gw.SetConfig(gwConfig)
+	gw.SetConfig(gwConfig)
 
 	cli.Init(VERSION, confPaths)
 
-	err = s.Gw.initialiseSystem()
+	err = gw.initialiseSystem()
 	if err != nil {
 		panic(err)
 	}
 
-	if s.Gw.GetConfig().EnableAnalytics && s.Gw.Analytics.GeoIPDB == nil {
+	if gw.GetConfig().EnableAnalytics && gw.Analytics.GeoIPDB == nil {
 		panic("GeoIPDB was not initialized")
 	}
 
-	configs := s.Gw.GetConfig()
+	configs := gw.GetConfig()
 
-	go s.Gw.RedisController.ConnectToRedis(ctx, func() {
-		if s.Gw.OnConnect != nil {
-			s.Gw.OnConnect()
+	go gw.RedisController.ConnectToRedis(s.ctx, func() {
+		if gw.OnConnect != nil {
+			gw.OnConnect()
 		}
 	}, &configs)
 
-	for {
-		if s.Gw.RedisController.Connected() {
-			break
-		}
-		time.Sleep(10 * time.Millisecond)
-	}
+	timeout, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
 
-	if !s.config.SkipEmptyRedis {
-		if err := s.emptyRedis(); err != nil {
-			panic(err)
-		}
+	connected := gw.RedisController.WaitConnect(timeout)
+	if !connected {
+		panic("can't connect to redis, timeout")
 	}
 
 	// Start listening for reload messages
-	if !s.Gw.GetConfig().SuppressRedisSignalReload {
-		go s.Gw.startPubSubLoop()
+	if !gw.GetConfig().SuppressRedisSignalReload {
+		go gw.startPubSubLoop()
 	}
 
-	if slaveOptions := s.Gw.GetConfig().SlaveOptions; slaveOptions.UseRPC {
+	if slaveOptions := gw.GetConfig().SlaveOptions; slaveOptions.UseRPC {
 		mainLog.Debug("Starting RPC reload listener")
-		s.Gw.RPCListener = RPCStorageHandler{
+		gw.RPCListener = RPCStorageHandler{
 			KeyPrefix:        "rpc.listener.",
 			SuppressRegister: true,
-			Gw:               s.Gw,
+			Gw:               gw,
 		}
 
-		s.Gw.RPCListener.Connect()
-		go s.Gw.rpcReloadLoop(slaveOptions.RPCKey)
-		go s.Gw.RPCListener.StartRPCKeepaliveWatcher()
-		go s.Gw.RPCListener.StartRPCLoopCheck(slaveOptions.RPCKey)
+		gw.RPCListener.Connect()
+		go gw.rpcReloadLoop(slaveOptions.RPCKey)
+		go gw.RPCListener.StartRPCKeepaliveWatcher()
+		go gw.RPCListener.StartRPCLoopCheck(slaveOptions.RPCKey)
 
-		go s.Gw.reloadLoop(time.Tick(time.Second))
```

Comment:
Leak, added background context cancellation check instead to cleanly stop the ticker.